tools/codegen/api/lazy.py

```diff
@@ -58,19 +63,36 @@ def process_ir_type(typ: Type) -> Union[BaseCType, VectorCType, OptionalCType, L
         raise AssertionError(f"unrecognized type {repr(typ)}")
 
 
-def isValueType(typ: Union[Type, BaseCType, OptionalCType, ConstRefCType, MutRefCType,
-                           ListCType, ArrayRefCType, ArrayCType, VectorCType, TupleCType]) -> bool:
+def isValueType(typ: CType) -> bool:
     """
     Given a type, determine if it is a Value-like type.  This is equivalent to
     being Tensor-like, but assumes the type has already been transformed.
     """
     if isinstance(typ, BaseCType):
-        return typ.type == valueT
+        # I am regretting my naming conventions, but now we are wrapping at::scalar in
+        # lazy value, while preserving other 'scalar' types as scalars in the IR
+        return typ.type == valueT or typ.type == scalarT
```

Comment:
I don't totally follow from the comment - can you give a code example?

It also doesn't seem like this really matches the description of the function: "This is equivalent to being Tensor-like", if we directly return true on scalar types.

Reply:
lazy tensor tracing promotes at::Scalars to become lazy tensors.  This enables treating the scalars as inputs to the lazy trace and not burning them into the program.

it could use a bit of fresh paint.  i think i might want to refactor some of this stuff but probably in a separate PR.  I could use your suggestions for that too more generally.   but I don't think I want to make too many changes at this PR since it's aiming to just sync up master to what we are already using on staging.  Better to develop the new version on staging where we can test it directly.  (until we finally finish merging enough to master that we can build/run entirely on master's CI.)